jupyter_server/services/contents/fileidmanager.py

```diff
@@ -0,0 +1,126 @@
+import os
+import sqlite3
+
+from jupyter_core.paths import jupyter_data_dir
+from traitlets import Unicode
+from traitlets.config.configurable import LoggingConfigurable
+
+
+class FileIdManager(LoggingConfigurable):
+    db_path = Unicode(
+        default_value=os.path.join(jupyter_data_dir(), "file_id_manager.db"),
+        help=(
+            "The path of the DB file used by `FileIdManager`. "
+            "Defaults to `jupyter_data_dir()/file_id_manager.db`."
+        ),
+        config=True,
+    )
+
+    def __init__(self, *args, **kwargs):
+        # pass args and kwargs to parent Configurable
+        super().__init__(*args, **kwargs)
+        # initialize connection with db
+        self.con = sqlite3.connect(self.db_path)
+        self.log.debug("Creating File ID tables and indices")
+        self.con.execute(
+            "CREATE TABLE IF NOT EXISTS Files(id INTEGER PRIMARY KEY, path TEXT NOT NULL UNIQUE)"
+        )
+        self.con.execute("CREATE INDEX IF NOT EXISTS ix_Files_path ON FILES (path)")
+        self.con.commit()
+
+    def _normalize_path(self, path):
+        """Normalizes a given file path."""
+        path = os.path.normcase(path)
+        path = os.path.normpath(path)
+        return path
+
+    def index(self, path):
+        """Adds the file path to the Files table, then returns the file ID. If
+        the file is already indexed, the file ID is immediately returned."""
+        path = self._normalize_path(path)
+        existing_id = self.get_id(path)
+        if existing_id is not None:
+            return existing_id
+
+        cursor = self.con.execute("INSERT INTO Files (path) VALUES (?)", (path,))
+        self.con.commit()
+        return cursor.lastrowid
+
+    def get_id(self, path):
+        """Retrieves the file ID associated with a file path. Returns None if
+        the file path has not yet been indexed."""
+        path = self._normalize_path(path)
+        row = self.con.execute("SELECT id FROM Files WHERE path = ?", (path,)).fetchone()
+        self.con.commit()
+        return row[0] if row else None
+
+    def get_path(self, id):
+        """Retrieves the file path associated with a file ID. Returns None if
+        the ID does not exist in the Files table."""
+        row = self.con.execute("SELECT path FROM Files WHERE id = ?", (id,)).fetchone()
+        self.con.commit()
```

Comment:
Thanks for the feedback regarding the control flow and making sure we only commit once for performance. All public methods should now only call a single commit before returning. `get_id()` and `get_path()` also need to call `commit()`, since they may actually update the DB if they detect an out-of-band move. Check out the implementation doc here for more detail: https://github.com/jupyter-server/jupyter_server/issues/940